lib/middleware/operation.js

```diff
@@ -32,55 +36,92 @@ function findPropertyOperations ({ types, method, term, resource }) {
+    if (moreOperations.length) {
+      return moreOperations.map(operation => ({ resource, operation }))
     }
 
-    const [operation, ...rest] = (operations || []).toArray()
+    return [{ resource, operation: null }]
+  })
+}
+
+async function invokeOperation (req, res, next) {
+  const { api } = req.hydra
+
+  if (!req.hydra.operation) {
+    const operationMap = mapOperations({ api, res })
+    const allowedMethods = new Set(operationMap
+      .filter(({ operation }) => operation)
+      .flatMap(({ operation }) => operation.out(ns.hydra.method).values))
 
-    if (!operation) {
+    if (!allowedMethods.size) {
       warn('no operations found')
-      let allowedOperations
-      if (req.hydra.term.equals(req.hydra.resource.term)) {
-        allowedOperations = findClassOperations(types)
-      } else {
-        allowedOperations = findPropertyOperations({ types, term: req.hydra.term, resource: req.hydra.resource })
+      if (operationMap.every(({ resource }) => 'property' in resource)) {
+        return next(new createError.NotFound())
       }
-
-      const allowedMethods = new Set(allowedOperations.out(ns.hydra.method).values)
-      res.setHeader('Allow', [...allowedMethods])
       return next(new createError.MethodNotAllowed())
     }
 
-    if (rest.length > 0) {
-      error('Multiple operations found')
-      return next(new Error('Ambiguous operation'))
+    res.setHeader('Allow', [...allowedMethods])
+    return next(new createError.MethodNotAllowed())
+  }
+
+  const handler = await api.loaderRegistry.load(req.hydra.operation.out(code.implementedBy), { basePath: api.codePath })
+
+  if (!handler) {
+    return next(new Error('Failed to load operation'))
+  }
+
+  handler(req, res, next)
+}
+
+function findOperation (operationMiddleware) {
+  return async function (req, res, next) {
+    const method = req.method === 'HEAD' ? 'GET' : req.method
+
+    let operations = mapOperations({ api: req.hydra.api, res, method })
+      .filter(({ operation }) => operation)
+
+    if (operationMiddleware) {
+      operations = operationMiddleware(operations, { req, res })
     }
 
-    const handler = await api.loaderRegistry.load(operation.out(code.implementedBy), { basePath: api.codePath })
+    if (!operations.length) {
+      return next()
+    }
 
-    if (!handler) {
-      return next(new Error('Failed to load operation'))
+    if (operations.length > 1) {
+      error('Multiple operations found')
+      return next(new Error('Ambiguous operation'))
     }
 
+    const [{ resource, operation }] = operations
+    req.hydra.resource = resource
     req.hydra.operation = operation
-    handler(req, res, next)
+    return next()
   }
 }
 
+function factory (middleware = {}) {
+  const router = Router()
+
+  router.use(findOperation(middleware.operations))
+
+  if (middleware.resource) {
+    router.use(middleware.resource)
+  }
+  router.use(invokeOperation)
```

Comment:
I'm a little concerned about this part. The `middleware.resource` has to be aware that the `req.hydra.resource` might be undefined if no resource+operation pair was selected.

It would make sense to move it inside `invokeOperation` handler but because it is also an express middleware then the result might be a little awkward.